### DISCUSSION TITLE
Mixing static library with dynamic

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#16864] [Q&A]
When i do the following I got an error:
1. `emmake make` create a static module - `someModule.a `
2. ` a.wasm`   linked with -s SIDE_MODULE.

Now, when `main.wasm` linked with -s MAIN_MODULE and with `someModule.a` and `a.wasm` 
I get a duplicate symbol error because `someModule.a`  and `main.wasm `has both probably linked with `libc`.
but, as far I understand,  SIDE_MODULE cannot be declared to   static library -` someModule.a` .
so , how it can be resolved if I need to keep the above modules structure  - 2 dynamic modules and one static?

**Comment by sbc100** [comment]
It should be fine to link static libraries into your main module.  Static libraries such as `someModule.a` cannot contain libc since libc only gets linked in the main module.

What symbols are you seeing duplicated?  Can you show the full failing link command and its output?

**Comment by aharondavid** [comment]
yes, here you go:

>emcc : warning : -sMAIN_MODULE + pthreads is experimental [-Wexperimental]
2>emcc : warning : USE_PTHREADS + ALLOW_MEMORY_GROWTH may run non-wasm code slowly, see https://github.com/WebAssembly/design/issues/1271 [-Wpthreads-mem-growth]
2>wasm-ld : error : duplicate symbol: fmemopen
2>>>> defined in C:/inetpub/wwwroot/SG/TestEmscripten/Core.Emscripten.wasm(sl_globals.o)
2>>>> defined in C:\emsdk\upstream\emscripten\cache\sysroot\lib\wasm32-emscripten\pic\libc-mt-debug.a(fmemopen.o)

>wasm-ld : error : duplicate symbol: mkdir
2>>>> defined in C:/inetpub/wwwroot/SG/TestEmscripten/Core.Emscripten.wasm(sl_core_emscripten.o)
2>>>> defined in C:\emsdk\upstream\emscripten\cache\sysroot\lib\wasm32-emscripten\pic\libc-mt-debug.a(mkdir.o)
2>

2>wasm-ld : error : duplicate symbol: swscanf
2>>>> defined in C:/inetpub/wwwroot/SG/TestEmscripten/Core.Emscripten.wasm(sl_core_emscripten.o)
2>>>> defined in C:\emsdk\upstream\emscripten\cache\sysroot\lib\wasm32-emscripten\pic\libc-mt-debug.a(swscanf.o)
2>

2>wasm-ld : error : duplicate symbol: wcsstr
2>>>> defined in C:/inetpub/wwwroot/SG/TestEmscripten/Core.Emscripten.wasm(sl_core_emscripten.o)
2>>>> defined in C:\emsdk\upstream\emscripten\cache\sysroot\lib\wasm32-emscripten\pic\libc-mt-debug.a(wcsstr.o)
2>

**Reply by sbc100** [reply]
The problem here is that `sl_core_emscripten.o` defined a bunch of libc symbols.. do you know why that is?  Where does that object come from?  Do you know why it defines those symbols?

Also, what is `Core.Emscripten.wasm` using the .wasm extension?    From the error message it looks like actually `.a` archive created by emar, so should probably be called `.a`.    

Also, I still don't see the command line that failed.. in order to help you the most it would be useful to see that.  (if you are using cmake you can build with `make VERBOSE=1` .. or just use `ninja`, if you want to see the full failing command line).

**Comment by aharondavid** [comment]
Here is the command line of :
`C:\emsdk\upstream\emscripten\\emcc.bat Emscripten/Debug/HttpNetFile.o Emscripten/Debug/IOFILE.o Emscripten/Debug/Source.o C:/Projects/simple_opengl-master/TheMainProject/libs/libxml2.a C:/Projects/simple_opengl-master/TheMainProject/libs/libgdal.a C:/Projects/simple_opengl-master/TheMainProject/libs/libproj.a C:/inetpub/wwwroot/SG/TestEmscripten/Core.Emscripten.wasm C:/inetpub/wwwroot/SG/TestEmscripten/SimpleProject.wasm -fexceptions -s WASM_BIGINT  -s FETCH=1 -g  --preload-file com.skyline.terraexplorer   --bind -fexceptions   -s USE_PTHREADS   -s PTHREAD_POOL_SIZE=navigator.hardwareConcurrency  -s ALLOW_MEMORY_GROWTH=1   -s MAIN_MODULE -O0 -o C:/inetpub/wwwroot/SG/TestEmscripten/TheMainProject.js`

I build my project by Emscripten Plugin for Visual Studio, and it is not expose the command line that build the Core.Emscripten.wasm above, but here is the flags that passed to the emcc:
```
AdditionalOptions=-Wsign-compare -Wno-import -Wformat=2 -Wmain -Wmissing-braces -w -Wformat-nonliteral -Wformat-nonliteral  -fpermissive -pthread   
PreprocessorDefinitions=__EMSCRIPTEN_PTHREADS__;_MB_CAPABLE;UNICODE;_UNICODE;_LIBCPP_HAS_NO_OFF_T_FUNCTIONS;HAVE_LONG_LONG;
```
and flags to the linker:  --bind -s SIDE_MODULE

I hope that it help
